packages/template-retail-react-app/app/commerce-api/utils.js

```diff
@@ -271,3 +272,23 @@ export const convertSnakeCaseToSentenceCase = (text) => {
  * Usually used as default for event handlers.
  */
 export const noop = () => {}
+
+/**
+ * WARNING: This function is relevant to be used in Hybrid deployments only.
+ * Compares the refresh_token keys for guest(cc-nx-g) and registered('cc-nx') login from the cookie received from SFRA with the copy stored in localstorage on PWA Kit
+ * to determine if the login state of the shopper on SFRA site has changed.
+ * @param {Storage} storage Cookie storage on PWA Kit in hybrid deployment.
+ * @param {LocalStorage} storageCopy Local storage holding the copy of the refresh_token in hybrid deployment.
+ * @returns {boolean} true if the keys do not match (login state changed), false otherwise.
+ */
+export function hasSFRAAuthStateChanged(storage, storageCopy) {
+    let refreshTokenKey =
+        (storage.get(refreshTokenGuestStorageKey) && refreshTokenGuestStorageKey) ||
+        (storage.get(refreshTokenRegisteredStorageKey) && refreshTokenRegisteredStorageKey)
+
+    let refreshTokenCopyKey =
+        (storageCopy.get(refreshTokenGuestStorageKey) && refreshTokenGuestStorageKey) ||
+        (storageCopy.get(refreshTokenRegisteredStorageKey) && refreshTokenRegisteredStorageKey)
+
+    return refreshTokenKey !== refreshTokenCopyKey
```

Comment:
what happens if user stays on PWA kit and keeps refreshing the page? (no SFRA)

It seems to me `hasSFRAAuthStateChanged` always evaluate to `true` and PWA will always use refresh token, and never re-use access token?

Reply:
if cookie doesn't exist, this function should return true

Reply:
also, let's add a test case for when refresh cookies does not exist.

Reply:
> what happens if user stays on PWA kit and keeps refreshing the page? (no SFRA)
I'm guessing by no SFRA you mean in case of a PWA Kit only setup:
In this case, the `storage` and `storageCopy` would both be pointing to `LocalStorage` and `refreshTokenKey` === `refreshTokenKeyCopy` so `hasSFRAAuthStateChanged` will be `false`

If by no SFRA you mean Hybrid Setup but the user hasn't yet navigated to SFRA,
Then `storage` will point to CookieStorage and `storageCopy` will point to LocalStorage in which case PWA Kit itself will write the refresh_token in both places. And if the user keeps refreshing on PWA kit without navigating to SFRA again `refreshTokenKey` === `refreshTokenKeyCopy` so `hasSFRAAuthStateChanged` will be `false`

Reply:
Sorry to be a stickler but why are we only comparing the keys and not the values of the refresh token?

By comparing just the keys we open ourselves to the following edge case in hybrid:

1. User signs in as a guest via SFRA
2. User goes to PWA - this sets an access token PWA side
3. User adds something to their cart via PWA
4. User goes back to SFRA, logs in (gets a cc-nx) then logs out (goes back to cc-nx-g)
5. User goes back to PWA

In the above, SFRA has changed the login state a couple of times but settled on a cc-nx-g when transitioning to PWA. If we compare only the keys, in this case we are not discarding the old access token.

Reply:
Updated `hasSFRALoginStateChanged(...)` to compare values if refershToken keys are same.

Thanks @vcua-mobify for catching that 🎉